osu.Game/Screens/Select/FilterQueryParser.cs

```diff
@@ -312,11 +323,45 @@ private static bool tryUpdateCriteriaRange<T>(ref FilterCriteria.OptionalRange<T
 
         private static bool tryUpdateLengthRange(FilterCriteria criteria, Operator op, string val)
         {
-            if (!tryParseDoubleWithPoint(val.TrimEnd('m', 's', 'h'), out double length))
+            List<string> parts = new List<string>();
+            GroupCollection groups = null;
+
+            if (
+                tryMatchRegex(val, @"^((?<hours>\d+):)?(?<minutes>\d+):(?<seconds>\d+)$", ref groups) ||
+                tryMatchRegex(val, @"^((?<hours>\d+(\.\d+)?)h)?((?<minutes>\d+(\.\d+)?)m)?((?<seconds>\d+(\.\d+)?)s)?$", ref groups) ||
+                tryMatchRegex(val, @"^(?<seconds>\d+(\.\d+)?)$", ref groups)
+            )
+            {
+                if (groups["seconds"].Success)
+                    parts.Add(groups["seconds"].Value + "s");
+                if (groups["minutes"].Success)
+                    parts.Add(groups["minutes"].Value + "m");
+                if (groups["hours"].Success)
+                    parts.Add(groups["hours"].Value + "h");
```

Comment:
I'm not sure why this is assembling strings anew, even though the input has just been parsed into recognisable groups. This shouldn't be required. The input is already recognised, use the matches directly.

Reply:
Reassembling them into an array makes them easier to check. Each of them is conditional (in other words every part can exist or not). And we should check if they satisfy these rules (assume they are ordered in any subsequence of `s, m, h`):
+ Each part are numbers (int or decimal) (already checked in the regexes)
+ All of them should be less than 60 except the last part
+ All of them shouldn't contain a decimal dot except the first part
+ Determine the minimum scale

If we assemble them into a sequence, all these checks are easy to do. But if we use the match directly, it should take us a lot of codes to check them.

Don't know if there is a better way exists, so I suppose to keep the reassembling method for now.